### PR TITLE
cwltool: init at 3.1.20210628163208

### PR DESCRIPTION
**Pull request by veprbl**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 130903` at 5ba30bbe run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built successfully:</summary>
  <ul>
    <li>cwltool</li>
    <li>python38Packages.bagit</li>
    <li>python38Packages.schema-salad</li>
    <li>python38Packages.shellescape</li>
    <li>python39Packages.bagit</li>
    <li>python39Packages.schema-salad</li>
    <li>python39Packages.shellescape</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/shellescape/default.nix:18:0:
```
   |
18 |     description = "Shell escape a string to safely use it as a token in a shell command (backport of Python shlex.quote)";
   | ^
```
</li>
  </ul>
</details>

File: pkgs/applications/science/misc/cwltool/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    mypy-extensions
+    prov
+    psutil
+    pydot
+    schema-salad
+    shellescape
+    typing-extensions
+  ];
+
+  doCheck = false; # hard to setup
```
Comment (SuperSandro2000) [inline pkgs/applications/science/misc/cwltool/default.nix:32]: Please explain why.

Can we add an pythonImportsCheck instead?

File: pkgs/development/python-modules/bagit/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+buildPythonPackage rec {
+  pname = "bagit";
+  version = "1.8.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "37df1330d2e8640c8dee8ab6d0073ac701f0614d25f5252f9e05263409cee60c";
+  };
+
+  buildInputs = [ setuptools-scm ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bagit/default.nix:16]:
```suggestion
  nativeBuildInputs = [ setuptools-scm ];
```

File: pkgs/development/python-modules/bagit/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+  doCheck = false; # tests are missing on Pypi
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bagit/default.nix:18]: Can we fetch from github instead?

File: pkgs/development/python-modules/shellescape/default.nix [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+buildPythonPackage rec {
+  pname = "shellescape";
+  version = "3.8.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "40b310b30479be771bf3ab28bd8d40753778488bd46ea0969ba0b35038c3ec26";
+  };
+
+  doCheck = false; # no tests on PyPI
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/shellescape/default.nix:15]: Can we fetch from github instead?

File: pkgs/applications/science/misc/cwltool/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/applications/science/misc/cwltool/default.nix:32]: I don't have an explanation, but feel free to add your own

**Comment by Artturin** [comment]
working tests
```patch
diff --git a/pkgs/applications/science/misc/cwltool/default.nix b/pkgs/applications/science/misc/cwltool/default.nix
index 21e3c47b323..c2f115a137b 100644
--- a/pkgs/applications/science/misc/cwltool/default.nix
+++ b/pkgs/applications/science/misc/cwltool/default.nix
@@ -1,14 +1,19 @@
 { lib
 , python3Packages
+, fetchFromGitHub
+, nodejs
+, git
 }:
 
 python3Packages.buildPythonApplication rec {
   pname = "cwltool";
   version = "3.1.20210628163208";
 
-  src = python3Packages.fetchPypi {
-    inherit pname version;
-    sha256 = "21b885f725420413d2f87eadc5e81c08a9c91beceda89b35d1a702ec4df47e52";
+  src = fetchFromGitHub {
+    owner = "common-workflow-language";
+    repo = "cwltool";
+    rev = version;
+    sha256 = "1m75m7fjya3qj3p2a764djf9qd8bgw0xlnlq539vr214mg7095x5";
   };
 
   postPatch = ''
@@ -16,6 +21,11 @@ python3Packages.buildPythonApplication rec {
       --replace 'prov == 1.5.1' 'prov'
   '';
 
+  # running egg_info error: [Errno 2] No such file or directory: 'git'
+  nativeBuildInputs = [ git ];
+  # Is this a runtime dep?
+  buildInputs = [ nodejs ];
+
   propagatedBuildInputs = with python3Packages; [
     argcomplete
     bagit
@@ -29,8 +39,35 @@ python3Packages.buildPythonApplication rec {
     typing-extensions
   ];
 
-  doCheck = false; # hard to setup
-  pythonImportsCheck = [ "cwltool" ];
+  checkInputs = with python3Packages; [
+    pytestCheckHook
+    pytest-mock
+    pytest-xdist
+    pytest-cov
+    rdflib-jsonld
+    # cwltool requires Node.js engine to evaluate and validate Javascript expressions
+    nodejs
+  ];
+
+  disabledTests = [
+    # network
+    "test_http_path_mapping"
+    # network
+    "test_content_types"
+    # network
+    "test_load_graph_fragment_from_packed"
+    # /build/source/cwltool/sandboxjs.py:290: FutureWarning: Possible nested set at position 2
+    "test_js_console_cmd_line_tool"
+    # contains getTrueShellExeName | AssertionError: assert b''
+    "test_env_filtering"
+  ];
+
+  disabledTestPaths = [
+    # ModuleNotFoundError: No module named 'arcp'
+    "tests/test_provenance.py"
+    # network
+    "tests/test_udocker.py"
+  ];
 
   meta = with lib; {
     homepage = "https://www.commonwl.org";
```

**Comment by SuperSandro2000** [comment]
Please disable pytest-cov.

**Pull request merged**